jlm/llvm/ir/operators/lambda.cpp

```diff
@@ -41,127 +41,96 @@ operation::copy() const
+  {
+    results.push_back(subregion()->result(n));
+  }
+  return results;
 }
 
-node::fctresult_range
-node::fctresults()
+[[nodiscard]] node::ContextVar
+node::MapInputContextVar(const rvsdg::input & input) const noexcept
 {
-  fctresiterator end(nullptr);
-
-  if (nfctresults() == 0)
-    return { end, end };
-
-  fctresiterator begin(fctresult(0));
-  return { begin, end };
+  JLM_ASSERT(rvsdg::TryGetOwnerNode<node>(input) == this);
+  return ContextVar{ const_cast<rvsdg::input *>(&input),
+                     subregion()->argument(operation().Type()->NumArguments() + input.index()) };
 }
 
-node::fctresult_constrange
-node::fctresults() const
+[[nodiscard]] std::optional<node::ContextVar>
+node::MapBinderContextVar(const rvsdg::output & output) const noexcept
 {
-  fctresconstiterator end(nullptr);
-
-  if (nfctresults() == 0)
-    return { end, end };
-
-  fctresconstiterator begin(fctresult(0));
-  return { begin, end };
+  JLM_ASSERT(rvsdg::TryGetOwnerRegion(output) == subregion());
+  auto numArguments = operation().Type()->NumArguments();
+  if (output.index() >= numArguments)
+  {
+    return ContextVar{ input(output.index() - operation().Type()->NumArguments()),
+                       const_cast<rvsdg::output *>(&output) };
+  }
+  else
+  {
+    return std::nullopt;
+  }
 }
 
-cvinput *
-node::input(size_t n) const noexcept
+[[nodiscard]] std::vector<node::ContextVar>
+node::GetContextVars() const noexcept
 {
-  return util::AssertedCast<cvinput>(structural_node::input(n));
+  std::vector<ContextVar> vars;
+  for (size_t n = 0; n < ninputs(); ++n)
+  {
+    vars.push_back(
+        ContextVar{ input(n), subregion()->argument(n + operation().Type()->NumArguments()) });
+  }
+  return vars;
 }
 
-lambda::output *
+rvsdg::output *
 node::output() const noexcept
 {
-  return util::AssertedCast<lambda::output>(structural_node::output(0));
-}
-
-lambda::fctargument *
-node::fctargument(size_t n) const noexcept
-{
-  return util::AssertedCast<lambda::fctargument>(subregion()->argument(n));
+  return structural_node::output(0);
 }
 
-lambda::cvargument *
-node::cvargument(size_t n) const noexcept
+node::ContextVar
+node::AddContextVar(jlm::rvsdg::output * origin)
 {
-  return input(n)->argument();
+  auto input = rvsdg::structural_input::create(this, origin, origin->Type());
+  auto argument = rvsdg::RegionArgument::Create(subregion(), input, origin->Type());
+  return ContextVar{ input, argument };
 }
 
-lambda::result *
-node::fctresult(size_t n) const noexcept
-{
-  return util::AssertedCast<lambda::result>(subregion()->result(n));
-}
-
-cvargument *
-node::add_ctxvar(jlm::rvsdg::output * origin)
-{
-  auto input = cvinput::create(this, origin);
-  return cvargument::create(subregion(), input);
-}
-
-rvsdg::RegionArgument &
+rvsdg::output &
 node::GetMemoryStateRegionArgument() const noexcept
 {
-  auto argument = fctargument(nfctarguments() - 1);
+  auto argument = GetFunctionArguments().back();
```

Comment:
This is expensive, no? You build an entire vector just to take the last element. It is possible to directly address the element, no?

These methods are wrong in the first place (after the introduction of memory state encoding), but let's leave them for now.

Reply:
Yes it is more expensive than it could be, I am not sure "GetMemoryStateRegionArgument" should be part of the generic lambda API (it is a very specific thing corresponding to a very specific encoding). I would leave the inefficiency for the moment until moving lambda to generic rvsdg and deciding how to proceed with the encoding.